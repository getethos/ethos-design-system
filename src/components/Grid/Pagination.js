import React from 'react'
import usePagination from './usePagination'
import PropTypes from 'prop-types'
import styles from './Pagination.module.scss'

const Pagination = ({ fetchPageCallback, renderCallback }) => {
  if (!fetchPageCallback || !renderCallback) {
    throw Error(
      'Pagination requires a fetchPageCallback and renderCallback parameters'
    )
  }

  const { fetchPage, pagingState, getPaginationNumbers } = usePagination({
    fetchPage: fetchPageCallback,
  })

  return (
    <>
      {renderCallback(pagingState.data)}
      <nav aria-label="pagination" className={styles.pagination}>
        <button
          className={[
            styles.paginationButtons,
            styles.paginationButtonsLeft,
          ].join(' ')}
          onClick={() => fetchPage(1)}
          aria-label="Goto Page 1"
        >
          &laquo;
        </button>
        {getPaginationNumbers()}
        <button
          className={[
            styles.paginationButtons,
            styles.paginationButtonsRight,
          ].join(' ')}
          onClick={() => fetchPage(pagingState.total_pages)}
          aria-label={`Goto Page ${pagingState.total_pages}`}
        >
          &raquo;
        </button>
      </nav>
    </>
  )
}

Pagination.propTypes = {
  fetchPageCallback: PropTypes.func.isRequired,
  renderCallback: PropTypes.func.isRequired,
}
Pagination.displayName = 'Pagination'

export default Pagination
