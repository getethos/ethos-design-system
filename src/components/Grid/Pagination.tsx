import { memo } from 'react'
import { usePagination } from './usePagination'
const DEFAULT_DISPLAY_PAGES_COUNT = 6
type PaginationProps = {
  fetchPageCallback?: (number) => any
  currentPage: number
  pageCount: number
  displayedPagesCount?: number
}
/**
 * Pagination component for use with the data grid
 *
 * @public
 *
 * @param {object} props - Component Props
 * @prop {number} [props.currentPage] - The current page
 * @prop {number} [props.pageCount] - The total number of pages available for the collection of things
 * @prop {number} [props.displayedPagesCount] - Number of pages displayed at a time
 * @prop {function} [props.fetchPageCallback] - Callback that fires when the next page should be fetched
 *
 * @return {JSX.Element}
 */
export const Pagination = memo(
  ({
    currentPage,
    pageCount,
    displayedPagesCount = DEFAULT_DISPLAY_PAGES_COUNT,
    fetchPageCallback,
  }: PaginationProps): JSX.Element | null => {
    if (
      typeof currentPage !== 'number' ||
      typeof pageCount !== 'number' ||
      !fetchPageCallback
    ) {
      throw Error(
        'Pagination requires currentPage, pageCount, and fetchPageCallback parameters'
      )
    }
    const { conditionallyRenderPagingButtons } = usePagination({
      fetchPageCallback,
    })
    return conditionallyRenderPagingButtons(
      pageCount,
      currentPage,
      displayedPagesCount
    )
  }
)
Pagination.displayName = 'Pagination'
