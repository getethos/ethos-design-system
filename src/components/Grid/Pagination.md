Used in tandem with the DataGrid component:

```jsx
import React, { memo, useEffect, useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Grid } from '../Grid/Grid'
import { Row } from '../Grid/Row'
import { Column } from '../Grid/Column'
import { useGridSorting } from '../Grid/useGridSorting.js'
import styles from '../Grid/grid-example.module.scss'

const toCaps = (str) => {
  const capitalizedWords = str.split(' ').map((word) => {
    return `${word.charAt(0).toUpperCase()}${word.slice(1)}`
  })
  return capitalizedWords.join(' ')
}

const LeGrid = ({ rows, columns }) => {
  const {
    rowsRefs,
    columnRefs,
    sortedRows,
    compareBy,
    updateRowsRefs,
    getSortIcon,
  } = useGridSorting(rows, columns)

  const fetchPageData = async (ev) => {
    ev.preventDefault()
    const key = ev.currentTarget.dataset.key || ''
    const idx = columns.findIndex((c) => c.name === key)
    const fetchFunction = columns[idx].fetchFn
    if (fetchFunction) {
      const applications = await fetchFunction(1, columns[idx].name)
      updateRowsRefs(applications.items)
    }
  }

  useEffect(() => {
    if (rows.length) {
      updateRowsRefs(rows)
    }
  }, [rows])

  return (
    <Grid rowRefs={rowsRefs} columnRefs={columnRefs}>
      <Row key="headers" columnRefs={columnRefs[0]}>
        {columns.map((col, x) => {
          if (col.sortable) {
            return (
              <Column
                name={col.name}
                key={uuidv4()}
                columnRef={columnRefs[0][x]}
                flexBasis={col.flexBasis}
                header
                interactive
                sortable
              >
                {(active, columnRef) => (
                  <a
                    href="./#"
                    onClick={fetchPageData}
                    data-key={col.name}
                    tabIndex={active ? 0 : -1}
                    className={styles.iconContainer}
                    ref={columnRef}
                  >
                    {toCaps(col.labelCopy || col.name)}
                    {getSortIcon(col.name)}
                  </a>
                )}
              </Column>
            )
          }
          return (
            <Column
              name={col.name}
              key={uuidv4()}
              columnRef={columnRefs[0][x]}
              flexBasis={col.flexBasis}
              header
            >
              {(active) => (
                <span
                  data-key={col.name}
                  className={active ? 'active' : undefined}
                >
                  {toCaps(col.labelCopy || col.name)}
                </span>
              )}
            </Column>
          )
        })}
      </Row>
      {sortedRows.map((row, y) => (
        <Row key={uuidv4()} columnRefs={rowsRefs[y]}>
          {columns.map((col, x) => {
            if (col.interactive) {
              return (
                <Column
                  name={col.name}
                  key={uuidv4()}
                  columnRef={rowsRefs[y][x]}
                  flexBasis={col.flexBasis}
                  interactive
                >
                  {(active, columnRef) => (
                    <a href="./#" tabIndex={active ? 0 : -1} ref={columnRef}>
                      {row[col.name]}
                    </a>
                  )}
                </Column>
              )
            }
            return (
              <Column
                name={col.name}
                key={uuidv4()}
                columnRef={rowsRefs[y][x]}
                flexBasis={col.flexBasis}
              >
                {() => (
                  <div className={styles.iconContainer}>
                    {row[col.name]}{' '}
                    {col.name === 'Type' && row[col.name] === 'Food' && (
                      <FontAwesomeIcon
                        icon={['far', 'hamburger']}
                        className={styles.icon}
                      />
                    )}
                  </div>
                )}
              </Column>
            )
          })}
        </Row>
      ))}
    </Grid>
  )
}

const GridAndPagination = memo(() => {
  const columns = [
    {
      name: 'id',
      flexBasis: '15%',
    },
    {
      name: 'name',
      labelCopy: 'color name',
      interactive: true,
      sortable: true /* will use default sort function */,
      flexBasis: '30%',
    },
    {
      name: 'color',
      labelCopy: 'hex',
      interactive: true,
      sortable: true /* will use default sort function */,
      flexBasis: '30%',
    },
    {
      name: 'pantone_value',
      labelCopy: 'pantone value',
      flexBasis: '15%',
    },
  ]

  const [pagingState, setPagingState] = useState({
    currentPage: 1,
    pageCount: 0,
    rows: [],
  })

  const fetchPage = async (pageNumber) => {
    // Unfortunately, there is no sort or order params offered by reqres
    // https://github.com/benhowdle89/reqres/issues/32
    const response = await fetch(
      `https://reqres.in/api/cases?page=${pageNumber}`,
      {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      }
    )

    const data = await response.json()

    // -------- reqres -------- //
    // These transform reqres response to Nora's paginate api
    const items = data.data
    const pageCount = data.total_pages
    // -------- ends: reqres -------- //

    setPagingState({
      currentPage: pageNumber,
      pageCount: pageCount,
      rows: items,
    })
  }

  const renderGrid = () => {
    if (pagingState.rows.length) {
      return <LeGrid rows={pagingState.rows} columns={columns} />
    }
    return null
  }

  useEffect(() => {
    fetchPage(1)
  }, [])

  return (
    <>
      {renderGrid()}
      <Pagination
        currentPage={pagingState.currentPage}
        pageCount={pagingState.pageCount}
        displayedPagesCount={6}
        fetchPageCallback={fetchPage}
      />
    </>
  )
})

;<GridAndPagination />
```

** Example Pagination views **

Pagination view with all page numbers displayed

```jsx
<Pagination
  currentPage={2}
  pageCount={5}
  displayedPagesCount={6}
  fetchPageCallback={() => {}}
/>
```

Pagination view with total count more than display pages

```jsx
<Pagination
  currentPage={2}
  pageCount={10}
  displayedPagesCount={6}
  fetchPageCallback={() => {}}
/>
```

Pagination view with left ellipsis

```jsx
<Pagination
  currentPage={15}
  pageCount={20}
  displayedPagesCount={10}
  fetchPageCallback={() => {}}
/>
```

Pagination view with left and right ellipsis

```jsx
<Pagination
  currentPage={6}
  pageCount={20}
  displayedPagesCount={10}
  fetchPageCallback={() => {}}
/>
```
