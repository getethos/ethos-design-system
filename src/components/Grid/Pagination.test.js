import React from 'react'
import TestRenderer, { act } from 'react-test-renderer'

import { Pagination } from './Pagination'
import styles from './Pagination.module.scss'

const responseStub = {
  page: 1,
  itemCount: 2,
  total: 6,
  pageCount: 3,
  data: [{ id: 1 }, { id: 2 }],
}

describe('Pagination Component', () => {
  let tree
  let fetchFn
  let renderFn

  beforeEach(() => {
    fetchFn = jest.fn()
    renderFn = jest.fn()
    fetchFn.mockReturnValue(responseStub)
  })

  afterEach(() => {
    tree = null
    fetchFn = null
    renderFn = null
  })

  it('default rendering', async () => {
    await act(async () => {
      tree = TestRenderer.create(
        <Pagination fetchPageCallback={fetchFn} renderCallback={renderFn} />
      )
    })
    let snapShot = tree.toJSON()
    expect(renderFn).toHaveBeenCalled()
    expect(snapShot).toMatchSnapshot()
  })

  it('fetches', async () => {
    tree = TestRenderer.create(
      <Pagination fetchPageCallback={fetchFn} renderCallback={renderFn} />
    )
    const root = tree.root
    const buttons = root.findAllByType('button')
    const [first] = buttons

    await act(async () => {
      first.props.onClick()
    })

    expect(fetchFn).toHaveBeenCalled()
    expect(renderFn).toHaveBeenCalled()
  })

  it('has aria labels', async () => {
    await act(async () => {
      tree = TestRenderer.create(
        <Pagination fetchPageCallback={fetchFn} renderCallback={renderFn} />
      )
    })
    const root = tree.root
    const buttons = root.findAllByType('button')
    expect(buttons.length).toEqual(5)
    const first = buttons[0]
    expect(first.props['aria-label']).toEqual('Goto Page 1')
    const last = buttons[4]
    expect(last.props['aria-label']).toEqual('Goto Page 3')
  })

  it('active page reflect current page', async () => {
    await act(async () => {
      tree = TestRenderer.create(
        <Pagination fetchPageCallback={fetchFn} renderCallback={renderFn} />
      )
    })
    const root = tree.root
    const buttons = root.findAllByType('button')
    const firstPageNumberButton = buttons[1]
    expect(firstPageNumberButton.props.className).toContain(styles.active)
  })
})
