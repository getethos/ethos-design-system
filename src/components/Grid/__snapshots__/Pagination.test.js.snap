// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component default rendering 1`] = `
<nav
  aria-label="pagination"
  className="pagination"
>
  <button
    aria-label="Goto Page 1"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    «
  </button>
  <button
    aria-label="Goto Page 1"
    className="paginationButtons"
    onClick={[Function]}
  >
    1
  </button>
  <button
    aria-current="page"
    aria-label="Goto Page 2"
    className="paginationButtons active"
    onClick={[Function]}
  >
    2
  </button>
  <button
    aria-label="Goto Page 3"
    className="paginationButtons"
    onClick={[Function]}
  >
    3
  </button>
  <button
    aria-label="Goto Page 3"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    »
  </button>
</nav>
`;

exports[`Pagination Component rendering with end ellipsis 1`] = `
<nav
  aria-label="pagination"
  className="pagination"
>
  <button
    aria-label="Goto Page 4"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    «
  </button>
  <button
    aria-label="Goto Page 1"
    className="paginationButtons"
    onClick={[Function]}
  >
    1
  </button>
  <button
    aria-label="Goto Page 2"
    className="paginationButtons"
    onClick={[Function]}
  >
    2
  </button>
  <button
    aria-label="Goto Page 3"
    className="paginationButtons"
    onClick={[Function]}
  >
    3
  </button>
  <button
    aria-label="Goto Page 4"
    className="paginationButtons"
    onClick={[Function]}
  >
    4
  </button>
  <button
    aria-current="page"
    aria-label="Goto Page 5"
    className="paginationButtons active"
    onClick={[Function]}
  >
    5
  </button>
  <button
    aria-label="Goto Page 6"
    className="paginationButtons"
    onClick={[Function]}
  >
    6
  </button>
  <button
    aria-label="Goto Page 7"
    className="paginationButtons"
    onClick={[Function]}
  >
    7
  </button>
  <button
    aria-label="Goto Page 8"
    className="paginationButtons"
    onClick={[Function]}
  >
    8
  </button>
  <button
    aria-disabled={true}
    aria-label="More Pages"
    className="paginationButtons"
    onClick={[Function]}
  >
    ...
  </button>
  <button
    aria-label="Goto Page 20"
    className="paginationButtons"
    onClick={[Function]}
  >
    20
  </button>
  <button
    aria-label="Goto Page 6"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    »
  </button>
</nav>
`;

exports[`Pagination Component rendering with start and end ellipsis 1`] = `
<nav
  aria-label="pagination"
  className="pagination"
>
  <button
    aria-label="Goto Page 5"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    «
  </button>
  <button
    aria-label="Goto Page 1"
    className="paginationButtons"
    onClick={[Function]}
  >
    1
  </button>
  <button
    aria-disabled={true}
    aria-label="More Pages"
    className="paginationButtons"
    onClick={[Function]}
  >
    ...
  </button>
  <button
    aria-label="Goto Page 4"
    className="paginationButtons"
    onClick={[Function]}
  >
    4
  </button>
  <button
    aria-label="Goto Page 5"
    className="paginationButtons"
    onClick={[Function]}
  >
    5
  </button>
  <button
    aria-current="page"
    aria-label="Goto Page 6"
    className="paginationButtons active"
    onClick={[Function]}
  >
    6
  </button>
  <button
    aria-label="Goto Page 7"
    className="paginationButtons"
    onClick={[Function]}
  >
    7
  </button>
  <button
    aria-label="Goto Page 8"
    className="paginationButtons"
    onClick={[Function]}
  >
    8
  </button>
  <button
    aria-label="Goto Page 9"
    className="paginationButtons"
    onClick={[Function]}
  >
    9
  </button>
  <button
    aria-disabled={true}
    aria-label="More Pages"
    className="paginationButtons"
    onClick={[Function]}
  >
    ...
  </button>
  <button
    aria-label="Goto Page 20"
    className="paginationButtons"
    onClick={[Function]}
  >
    20
  </button>
  <button
    aria-label="Goto Page 7"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    »
  </button>
</nav>
`;

exports[`Pagination Component rendering with start ellipsis 1`] = `
<nav
  aria-label="pagination"
  className="pagination"
>
  <button
    aria-label="Goto Page 14"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    «
  </button>
  <button
    aria-label="Goto Page 1"
    className="paginationButtons"
    onClick={[Function]}
  >
    1
  </button>
  <button
    aria-disabled={true}
    aria-label="More Pages"
    className="paginationButtons"
    onClick={[Function]}
  >
    ...
  </button>
  <button
    aria-label="Goto Page 13"
    className="paginationButtons"
    onClick={[Function]}
  >
    13
  </button>
  <button
    aria-label="Goto Page 14"
    className="paginationButtons"
    onClick={[Function]}
  >
    14
  </button>
  <button
    aria-current="page"
    aria-label="Goto Page 15"
    className="paginationButtons active"
    onClick={[Function]}
  >
    15
  </button>
  <button
    aria-label="Goto Page 16"
    className="paginationButtons"
    onClick={[Function]}
  >
    16
  </button>
  <button
    aria-label="Goto Page 17"
    className="paginationButtons"
    onClick={[Function]}
  >
    17
  </button>
  <button
    aria-label="Goto Page 18"
    className="paginationButtons"
    onClick={[Function]}
  >
    18
  </button>
  <button
    aria-label="Goto Page 19"
    className="paginationButtons"
    onClick={[Function]}
  >
    19
  </button>
  <button
    aria-label="Goto Page 20"
    className="paginationButtons"
    onClick={[Function]}
  >
    20
  </button>
  <button
    aria-label="Goto Page 16"
    className="paginationButtons"
    disabled={false}
    onClick={[Function]}
  >
    »
  </button>
</nav>
`;
